### PR TITLE
multiple fixes related to counter-battery fire

### PR DESCRIPTION
**Pull request by NickAragua**
These prevent several game-blocking and non-game-blocking NPEs which occur when handling counter-battery fire against a unit that was previously destroyed.

Fixes mekhq #2733

### CODE REVIEW

**Comment by HammerGS** [comment]
This supposed to point to MekHQ 2733?

File: megamek/src/megamek/common/Targetable.java [hunk 1]
```diff
@@ -151,4 +151,17 @@ default boolean tracksHeat() {
     
     @Override
     int hashCode();
+    
+    /**
+     * Utility function used to safely tell whether two Targetables are in the same hex.
+     * Does not throw exceptions in case of nulls.
+     */
+    public static boolean atSamePosition(Targetable first, Targetable second) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/Targetable.java:159]: Nullable annotate both first and second

File: megamek/src/megamek/common/weapons/ArtilleryWeaponIndirectFireHandler.java [hunk 2]
```diff
@@ -396,7 +396,11 @@ else if ((null != bestSpotter) && !(this instanceof ArtilleryWeaponDirectFireHan
             AreaEffectHelper.clearMineFields(targetPos, Minefield.CLEAR_NUMBER_WEAPON, ae, vPhaseReport, game, server);
         }
 
-        if(aaa.getTarget(game).isOffBoard()) {
+        Targetable updatedTarget = aaa.getTarget(game);
+        
+        // the attack's target may have been destroyed or fled since the attack was generated
+        // so we need to carry out offboard/null checks against the "current" version of the target.
+        if((updatedTarget != null) && updatedTarget.isOffBoard()) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/weapons/ArtilleryWeaponIndirectFireHandler.java:403]:
```suggestion
        if ((updatedTarget != null) && updatedTarget.isOffBoard()) {
```

File: megamek/src/megamek/server/Server.java [hunk 3]
```diff
@@ -13481,14 +13481,16 @@ private void processAttack(Entity entity, Vector<EntityAction> vector) {
...
                 for (Enumeration<AttackHandler> j = game.getAttacks(); !firingAtNewHex
                         && j.hasMoreElements(); ) {
                     WeaponHandler wh = (WeaponHandler) j.nextElement();
                     if (wh.waa instanceof ArtilleryAttackAction) {
                         ArtilleryAttackAction oaaa = (ArtilleryAttackAction) wh.waa;
+                        
                         if ((oaaa.getEntityId() == aaa.getEntityId())
-                            && !oaaa.getTarget(game).getPosition()
-                                .equals(aaa.getTarget(game).getPosition())) {
+                            && !Targetable.atSamePosition(oaaa.getTarget(game), attackTarget)) {
```
Comment (Windchild292) [inline megamek/src/megamek/server/Server.java:13493]: Additional tab for spacing

**Comment by NickAragua** [comment]
Yes

File: megamek/src/megamek/common/Targetable.java (hunk 1 above)
Comment (Windchild292) [reply megamek/src/megamek/common/Targetable.java:159]: Further, for the name... isAtSamePosition would probably be a better name than atSamePosition.

**Review by Windchild292 — COMMENTED** [review]
Mostly looks good, just a few comments first

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**